src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java

```diff
@@ -835,15 +826,29 @@ private static DetailAST getFirstNodeInsideSwitchBlock(
     }
 
     /**
-     * Helper method for getFirstNodeInsideSwitchBlock to return the first CASE_GROUP or
-     * SWITCH_RULE ast.
+     * Helper method for getFirstNodeInsideSwitchBlock to return all variable
+     * usage expressions inside a given switch block.
      *
      * @param block the switch block to check.
-     * @return DetailAST of the first CASE_GROUP or SWITCH_RULE.
+     * @param variable variable which is checked for in switch block.
+     * @return List of usages or empty list if none are found.
      */
-    private static DetailAST getFirstCaseGroupOrSwitchRule(DetailAST block) {
-        return Optional.ofNullable(block.findFirstToken(TokenTypes.CASE_GROUP))
-            .orElseGet(() -> block.findFirstToken(TokenTypes.SWITCH_RULE));
+    private static List<DetailAST> getVariableUsageExpressionsInsideSwitchBlock(DetailAST block,
+                                                                            DetailAST variable) {
+        return Optional
+                .ofNullable(block.findFirstToken(TokenTypes.CASE_GROUP))
+                .or(() -> Optional.ofNullable(block.findFirstToken(TokenTypes.SWITCH_RULE)))
```

Comment:
Nested optional... this is not good. This return statement is complex and hard to read/ understand, and will be difficult to debug in the future.

Consider that we have three cases of interest to this check:
1. switch block has a CASE_GROUP child
2. switch block has a SWITCH_RULE child
3. switch block has no children that we care about

Then please do something like:

```java
        final Optional<DetailAST> b = TokenUtil.findFirstTokenByPredicate(b,
                c -> c.getType() == TokenTypes.SWITCH_RULE || c.getType() == TokenTypes.CASE_GROUP
        );
```

Now, we do not deal with null, and we have an optional that clearly represents scenario above.

Then:
1. proceed with list creation, we just return an empty list if we have no usages to register
2. if optional block child is present, register variable usages

suggestion:
```
        final List<DetailAST> variableUsageExpressions = new ArrayList<>();

        b.ifPresent(blockStatement -> {
            TokenUtil.forEachChild(block, blockStatement.getType(), child -> {
                final DetailAST lastNodeInCaseGroup = child.getLastChild();
                if (isChild(lastNodeInCaseGroup, variable)) {
                    variableUsageExpressions.add(lastNodeInCaseGroup);
                }
            });
        });

        return variableUsageExpressions;

```
@romani I am open to suggestions or discussion on this, but I do not think current implementation is easy to read/ reason about/ maintain

@stoyanK7 please read more about Optional type at https://stuartmarks.wordpress.com/2016/09/27/vjug24-session-on-optional/ if you are interested, I found this article/presentation last week and found it valuable.

Reply:
I am ok with proposal.
I am not sure if it is much better, but I. Not firm on this point.

Reply:
@stoyanK7, please do update.

Reply:
Done.

Took the time to watch the 50 minute presentation. I have mixed feelings about this now, at the same time they say single `Optional` is okay but then not every null should use `Optional`/do not use it for simple null check. It is really dependent on situation.

As for me, both are okay here. No complaints.